docs/architecture/deep-dives/state/derived-state.md

```diff
@@ -0,0 +1,129 @@
+# Derived State
+
+It is common to need to cache the result of expensive work that does not represent true alterations
+in application state. Derived state exists to store this kind of data in memory and keep it up to
+date when the underlying observable state changes.
+
+## `DeriveDefinition`
+
+Derived state has all of the same issues with storage and retrieval that normal state does. Similar
+to `KeyDefinition`, Derived state depends on `DeriveDefinition`s to define magic string keys to
+store and retrieve data from a cache. Unlike normal state, Derived state is always stored in memory.
+It still takes a `StateDefinition`, but this is used only to define a namespace for the derived
+state, the storage location is ignored. _This can lead to collisions if you use the same key for two
+different derived state definitions in the same namespace._
+
+Derive definitions can be created in two ways:
+
+<a name="deriveDefinitionFactories"></a>
+
+```typescript
+new DeriveDefinition(STATE_DEFINITION, "uniqueKey", _DeriveOptions_);
+
+// or
+
+const keyDefinition: KeyDefinition<T>;
+DeriveDefinition.from(keyDefinition, _DeriveOptions_);
+```
+
+The first allows building from basic building blocks, the second recognizes that derived state is
+often built from existing state and allows you to create a definition from an existing
+`KeyDefinition`. The resulting `DeriveDefinition` will have the same state namespace, key, and
+`TFrom` type as the `KeyDefinition` it was built from.
+
+### Type Parameters
+
+`DeriveDefinition`s have three type parameters:
+
+- `TFrom`: The type of the state that the derived state is built from.
+- `TTo`: The type of the derived state.
+- `TDeps`: defines the dependencies required to derive the state. This is further discussed in
+  [Derive Definition Options](#derivedefinitionoptions).
+
+### `DeriveDefinitionOptions`
+
+[The `DeriveDefinition` section](#deriveDefinitionFactories) specifies a third parameter as
+`_DeriveOptions_`, which is used to fully specify the way to transform `TFrom` to `TTo`.
+
+- `deserializer` - For the same reasons as [Key Definition Options](#keydefinitionoptions),
+  `DeriveDefinition`s require have a `deserializer` function that is used to convert the stored data
+  back into the `TTo` type.
+- `derive` - A function that takes the current state and returns the derived state. This function
+  takes two parameters
+  - `from` - The latest value of the parent state.
+  - `deps` - dependencies used to instantiate the derived state. These are provided when the
+    `DerivedState` class is instantiated. This object should contain all of the application runtime
+    dependencies for transform the from parent state to the derived state.
+- `cleanupDelayMs` (optional) - Takes the number of milliseconds to wait before cleaning up the
```

Comment:
❓ What's the underlying purpose for this?

Reply:
This is effectively the lifetime of derived values to live in memory once observers unsubscribe.

If you have some data that is expensive to derive, you may want to increase this to minimize the chance that data must be re-calculated.